bpf/lib/nodeport.h

```diff
@@ -541,13 +542,16 @@ static __always_inline int nodeport_lb6(struct __ctx_buff *ctx,
 
 	case CT_ESTABLISHED:
 	case CT_REPLY:
+		if (ct_state.rev_nat_index != svc->rev_nat_index) {
+			goto redo_all;
+		}
```

Comment:
Small nit: no extra braces here and below (kernel style).

Reply:
Isn't this racy? E.g. what happens if we had an earlier entry in there where packets are in flight and replies to be expected to go out from the Pod to the client?

Reply:
Otoh, we have the same logic in https://github.com/cilium/cilium/blob/master/bpf/bpf_lxc.c#L199 . Given there can only be one such svc at the same time, it really must have been stale. Could we consolidate such check with the one from bpf_lxc and add a comment on why it's okay to redo the full tuple?

Reply:
> Isn't this racy?

The stale conntrack entry problem can only happen when a packet is sent from the same IP addr and port. So, this means that the client before sending a request to the other svc waited long enough (TIME_WAIT) until the port became available again. Therefore, I assume that such races are very unlikely to happen.

> Could we consolidate such check with the one from bpf_lxc into a small helper and add a comment on why it's okay to redo the full tuple?

The check in bpf_lxc.c compares rev_nat_index from two ct_states, while in nodeport.h - from svc and ct_state. Probably it's OK to replace the former to check against svc instead of ct_state_new, so then the consolidation would make sense.

> Also, is there any advantage to explicitly delete the existing one first?
ct_delete{4,6}() also tries to remove related SNAT entries. Perhaps, it's fine to keep the SNAT related entries in the case of the nodeport.h, as it can be reused by snat_process() later on. In the case of bpf_lxc.c, do you see a case in which SNAT entries would be created?

Reply:
Not that I'm aware of, so it would make sense to have both logic the same, meaning we likely could just remove the ct_delete*().

Reply:
PTAL - I removed the ct_delete*()'s, but I didn't add a helper "ct_belongs_to_svc()", as IMO it might confuse developers by allowing them to think that passing any CT entry would test whether it belongs to the given svc. This is a wrong assumption, as not all entries have rev_nat_index set (e.g. CT_INGRESS).